specs/VEP-Milestone25-vector-database-ingestion/README.md

```diff
@@ -0,0 +1,147 @@
+- [Alternatives](#alternatives)
+
+## Summary
+
+With the rise in popularity of LLMs and RAG we see VDK as a core component to getting the data where we need it to be.
+VDK's strengths are ETL tasks. We see that its very well suited to populating the databases needed for RAG.
+
+
+## Glossary
+
+LLM: Large language model. The most ubiquitous example of this is chatgpt. It is a specialized type of artificial intelligence (AI) that has been trained on vast amounts of text to understand existing content and generate original content.
+RAG: Retrival augmented generation. Additional information is passed to the LLM through the prompt. This additional information can help it generate better and more context aware responses.
+Vector database: A database which supports storing vectors(arrays of numbers) and doing similarity searches between vectors(cosine distance, dot product etc...).
+PGVector: A postgres extension which enables similarity searches in postgres and a vector datatype.
+
+
+## Motivation
+
+#### Example problem scenario:
+A company has a powerful private LLM chatbot.
+However they want it to be able to answer questions using the latest version of confluence docs/jira tickets etc...
+Retraining every night on the latest tickets/docs is not feasible.
+Instead, they opt to use RAG to improve the chatbot responses.
+
+This leaves them with the question.
+How do we populate the data?
+Steps they need to complete:
+1. Read data from confluence/jira
+2. Chunk into paragraphs(or something similar)
+3. Embed into vector space
+4. save Vector and paragraph in vector database
+5. remove old information. For example if we are scraping jira every hour and we are writing details to the vector database we need to make sure we clean up all embeddings/chunks which were generated from old versions of the ticket.
+
+We will build a datajob in VDK which reads data from confluence or jira and writes it to a postgres instance with PGVector enabled. A embedding model will be running on a different machine which will be exposed through an API.
+We will make requests to the API to create embeddings for us.
+
+After this datajob is running we will create a template from this in which we think customers will be able to adopt to meet their use cases.
+
+#### Benefits to customers:
+They will be able to follow our template to quickly create similar jobs. This gives customers the ability to get up and runnning quickly.
+
+## Requirements and goals
+1. There should be a single pipelines which given jira/confluence credentials can scrape the source
+2. it should chunk up the information, embed it and then save it
+3. The systems should be easily configurable
+   1. Read from different sources
+   2. Different chunks sizes
+   3. Different embedders
+   4. Extra columns saved in database
+4. There should be an example on how to build your own ingestion pipeline
+5. Only scraping new data and removing old data must be supported
```

Comment:
= What will be in place for efficient data cleaning, for example for outdated or redundant information?

- How does the system handle version control and data integrity issues? What if we have configured wrong "incremental" key for example. 

- You mentioned it before (in a meeting) but what about user permissions? What if user doesn't have access to part of confluence. It's fine to just mark this as Non goal at this moment since we are not ready to tackle it. 

- How is the overall user experience, particularly in terms of ease of setting up and monitoring the data pipeline. I think it would be useful to sketch a pseudo-code type of data job showing what user is doing (it needs not be accurate or final) 

- How does the framework would handle errors or interruptions in data flow

Reply:
**What will be in place for efficient data cleaning, for example for outdated or redundant information?**

This is the million dollar question, I don't know how this should be implemented. 
I think by implementing this well along with a few other features, that customers would look to use our framework. 

**How does the system handle version control and data integrity issues? What if we have configured wrong "incremental" key for example.**

However this is currently handled in VDK is how it should be handled here


**You mentioned it before (in a meeting) but what about user permissions? What if user doesn't have access to part of confluence. It's fine to just mark this as Non goal at this moment since we are not ready to tackle it.**

Metadata will be propagated through. so the user access can be implemented a retrieval time. 

**How is the overall user experience, particularly in terms of ease of setting up and monitoring the data pipeline. I think it would be useful to sketch a pseudo-code type of data job showing what user is doing (it needs not be accurate or final)**


I have provided pseudo-code. Do you want to take a look and then add more questions here. 



**How does the framework would handle errors or interruptions in data flow**

This should work the same way it is currently handled in VDK